frontend/src/hooks/useValidation.js

```diff
@@ -0,0 +1,341 @@
+        );
+      }
+    }
+
+    return true;
+  }
+
+  // Validates all date ranges for a dateable feature from `dates`
+  function validateFeatureDates(dateableId, datesObj = dates) {
+    const dateableFeature = datesObj[dateableId];
+
+    const dateTypeGroups = Object.values(dateableFeature);
+
+    // Clear all errors for this dateable feature:
+    // Build a list of all field IDs to clear from `errors`
+    const fieldIds = dateTypeGroups.flatMap((dateRanges) =>
+      dateRanges.flatMap((dateRange, index) => {
+        const dateRangeId = `${dateRange.dateableId}-${dateRange.dateType.id}-${index}`;
+
+        return [`start-date-${dateRangeId}`, `end-date-${dateRangeId}`];
+      }),
+    );
+
+    // Remove any errors for this dateable feature before revalidating
+    setErrors((prevErrors) => omit(prevErrors, fieldIds));
+
+    const dateTypeGroupsValid = dateTypeGroups.every((dateRanges) =>
+      // Loop over date ranges for the date type
+      dateRanges.every((dateRange, index) =>
+        validateDateRange({
+          dateRange,
+          start: dateRange.startDate,
+          startDateId: `start-date-${dateRange.dateableId}-${dateRange.dateType.id}-${index}`,
+          end: dateRange.endDate,
+          endDateId: `end-date-${dateRange.dateableId}-${dateRange.dateType.id}-${index}`,
+          datesObj,
+        }),
+      ),
+    );
+
+    return dateTypeGroupsValid;
+  }
+
+  function validateForm() {
+    // Clear errors from previous validation
+    setErrors({});
+
+    // Validate notes
+    if (!validateNotes()) return false;
+
+    // Validate all dates:
+    // Loop over dateable features from `dates` (campsite groupings)
+    const dateableIds = Object.keys(dates);
+    const validDates = dateableIds.every((dateableId) =>
+      validateFeatureDates(dateableId),
+    );
+
+    // @TODO: Scroll the first error into view
+
+    return validDates;
+  }
+
+  // Validates a date range, then its parent dateable feature
+  function onUpdateDateRange({
+    dateRange,
+    start,
+    startDateId,
+    end,
+    endDateId,
+    // Allow overriding dates during state updates
+    datesObj = dates,
+  }) {
+    // Don't validate until the form has been submitted
+    if (!formSubmitted) return true;
+
+    const { dateableId } = dateRange;
+
+    // Validate the date range that changed
+    const rangeValid = validateDateRange({
+      dateRange,
+      start,
+      startDateId,
+      end,
+      endDateId,
+      datesObj,
+    });
+
+    // If the changed dateRange is invalid, don't validate anything else
+    if (!rangeValid) return false;
+
+    // If the changed dateRange is valid, validate the whole campsite grouping feature
+    // to resolve any inter-dependent date range validations.
+
+    // @TODO: This unnecessarily validates the changed dateRange twice;
+    // Consider refactoring to validate once, but only show relevant errors.
```

Comment:
It could be easier to just call validateFeatureDates for every change to avoid calling validateDateRange twice. Realistically, each feature would have at most like 4 dates? So, it wouldn't be a big performance issue.
But, I'm not totally against the duplicate calls either if we return false as soon as we find an error in the dateRange being edited.

Reply:
Just updated it to validate the whole feature at once on change, instead of starting with just the one date range